--- a/init-boot.sh
+++ b/init-boot.sh
@@ -17,28 +17,31 @@
 wait=""
 fstype="auto"
 
-[ ! -f /etc/platform-preboot ] || . /etc/platform-preboot
+for bootarg in `cat /proc/cmdline`; do
+    case "$bootarg" in
+        root=*) rootdev="${bootarg##root=}" ;;
+        ro) opt="ro" ;;
+        rootwait) wait="yes" ;;
+        rootfstype=*) fstype="${bootarg##rootfstype=}" ;;
+    esac
+done
 
 if [ -z "$rootdev" ]; then
-    for bootarg in `cat /proc/cmdline`; do
-        case "$bootarg" in
-            root=*) rootdev="${bootarg##root=}" ;;
-            ro) opt="ro" ;;
-            rootwait) wait="yes" ;;
-            rootfstype=*) fstype="${bootarg##rootfstype=}" ;;
-        esac
-    done
+    echo "rootfs not specified in cmd args" > /dev/ttyTCU0
+    [ ! -f /etc/platform-preboot ] || . /etc/platform-preboot
+else
+    echo "rootfs from cmd args: ${rootdev}"> /dev/ttyTCU0
 fi
 
 if [ -n "$wait" -a ! -b "${rootdev}" ]; then
-    echo "Waiting for ${rootdev}..."
+    echo "Waiting for ${rootdev}..." > /dev/ttyTCU0
     while true; do
         test -b "${rootdev}" && break
         sleep 0.1
     done
 fi
 
-echo "Mounting ${rootdev}..."
+echo "Mounting ${rootdev}..." > /dev/ttyTCU0
 [ -d /mnt ] || mkdir -p /mnt
 count=0
 while [ $count -lt 5 ]; do
@@ -52,7 +55,7 @@
 
 [ ! -f /etc/platform-pre-switchroot ] || . /etc/platform-pre-switchroot
 
-echo "Switching to rootfs on ${rootdev}..."
+echo "Switching to rootfs on ${rootdev}..." > /dev/ttyTCU0
 mount --move /sys  /mnt/sys
 mount --move /proc /mnt/proc
 mount --move /dev  /mnt/dev
